### PR TITLE
✨🐛 support switching providers post-connect

### PR DESCRIPTION
**Pull request by arlimus**
In the original design, the connect call would always return the connected asset. However, we have some legacy code that behaved differently in v8. When it was translated to v9, we had to return the fully discovered asset, which is the asset that we actually want to connect.

For now, the connect call will continue to fulfill two functions:
1. discover the actual asset and
2. establish the connection

During this time, we will check once if the provider changed after the connect call and switch providers if that happened.

We want a distinct discover and connect call in the future.

Note:
1. The connect call on the shell behaves differently, happy to adjust it if we need it there too.
2. If additional providers get connected via resources, they do not go through this procedure. This is intentional. Example: The os provider calls out to the network provider. There is no way for that process to all of a sudden end up with a completely different provider (that would also be a nightmare to debug)

### CODE REVIEW

**Comment by arlimus** [comment]
Feel free to hold off on it if it doesn't solve the problem

**Comment by czunker** [comment]
I tried it with the ebs snapshot scan (#2171). It was only working with builtin providers. https://github.com/mondoohq/cnquery/pull/2127/files was working in both cases (builtin and plugin). Perhaps we can combine these two somehow. I've currently no clue, what's the problem, sorry.

**Comment by imilchev** [comment]
If this works for builtin providers but not for installed ones, then it's probably an issue with pointers that get set in the original asset when using builtin because we are changing the pointer in the same process. For installed providers that doesn't work since we adjust the asset over gRPC. That's just a guess based on stuff I have seen so far

**Comment by vjeffrey** [comment]
sorry it took me so long to get back to this, i kept trying to test in between other things and not paying enough attention (its working well against most machines, but the machine i was testing against seems to be having troubles, but that's unrelated to this)

with this change + one other small change in the ebs provider, things work! thank you!

**Comment by vjeffrey** [comment]
hmmm, i think something else changed since this first came up -- things seem to work without this change now

**Comment by arlimus** [comment]
Holding off on it for now.

**Comment by czunker** [comment]
Still needed for #2126 .

File: providers/runtime.go [hunk 1]
```diff
@@ -216,6 +224,32 @@ func (r *Runtime) Connect(req *plugin.ConnectReq) error {
...
+	//
+	// Additionally, we do not loop this connect+recheck approach indefinitely.
+	// We only run it once and only accept one asset switch. This will be
+	// changed once we have an explicit discover call in plugins.
+	postProvider, err := r.providerForAsset(r.Provider.Connection.Asset)
+	if err != nil {
+		return err
+	}
+	if postProvider.ID != r.Provider.Instance.ID {
+		req.Asset = r.Provider.Connection.Asset
```
Comment (czunker) [inline providers/runtime.go:245]: @arlimus I needed to add this line to make it also work with plugins. With this change, the results also looks like the one in v8.

I tried to set all this inside the GCP provider, but it does not get transferred to the os provider.

**Comment by czunker** [comment]
@vjeffrey Any objects to merge this? You mentioned EBS scanning is now working without this change.

@arlimus Do you agree with the additional change?

File: providers/runtime.go (hunk 1 above)
Comment (arlimus) [reply providers/runtime.go:245]: Makes sense, thank you for the catch!

**Review by czunker — APPROVED** [review]
Thanks @arlimus

**Pull request merged**